### PR TITLE
TypeLayouts

### PR DESCRIPTION
**Pull request by pranavtbhat**
I've created this basic graph representation that should be compatible with other Julia graph libraries such as LightGraphs.jl and Graphs.jl.  The representation is very light; other auxillary structures such as labels etc. should be passed into functions separately. 

The graph supports both Adjacency List and Adjacency Matrix (SparseMatrixCSC also). I don't anticipate using `redistribute` on the GraphLayout though.

### CODE REVIEW

File: src/graph-layouts.jl [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+###
+
+# Allowed Graph datastructures:
+typealias AdjacencyList Array{Array{Int, 1}, 1}
+typealias AdjacencyMatrix AbstractArray{Int, 2}                      # Covers sparse matrices
+typealias GraphStruct Union{AdjacencyList, AdjacencyMatrix}
+
+
+# Generic (undirected or directed, up to the user to decide) Graph
+type DistGraph{S<:GraphStruct}
```
Comment (shashi) [inline src/graph-layouts.jl]: why is this needed? `distribute` combined with  `GraphLayout` can do the job right?

File: src/graph-layouts.jl [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+###
+# Graph layouts
+###
+
+immutable GraphLayout{S<:GraphStruct} <: AbstractLayout end
+
+"""
+Given an input vertex range, slice it into sub-ranges.
+"""
+function slice(ctx, vrange::UnitRange{Int}, ::GraphLayout, targets)
```
Comment (shashi) [inline src/graph-layouts.jl]: This function should be named something else.

File: src/graph-layouts.jl [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+# Allowed Graph datastructures:
+typealias AdjacencyList Array{Array{Int, 1}, 1}
+typealias AdjacencyMatrix AbstractArray{Int, 2}                      # Covers sparse matrices
+typealias GraphStruct Union{AdjacencyList, AdjacencyMatrix}
+
+
+# Generic (undirected or directed, up to the user to decide) Graph
+type DistGraph{S<:GraphStruct}
+    nv::Int                            # not really required, but kept for convenience
+    vertices::UnitRange{Int}
```
Comment (shashi) [inline src/graph-layouts.jl]: So far I don't see why we need this field.

Am thinking of allowing `slice` to return some metadata about each chunk and then storing that metadata in the DistMemory. So that each layout type can put some metadata along with the chunks to facilitate things like `length`, `getindex

Comment (pranavtbhat) [inline src/graph-layouts.jl]: The vertices field denotes the range of vertices present locally. It allows the workers to determine the global index of the first worker they have. But I guess the metadata will do something similar.

File: src/graph-layouts.jl (hunk 2 above)
Comment (pranavtbhat) [inline src/graph-layouts.jl]: I'll change the name.

File: src/graph-layouts.jl (hunk 1 above)
Comment (pranavtbhat) [inline src/graph-layouts.jl]: This representation is similar to what Graphs.jl and LightGraphs.jl use. Implementing `convert` methods will allow users to apply algorithms directly on their graphs. Writing algorithms will also be more convenient with a representation such as this. 

Will this cause any performance problems?

**Comment by pranavtbhat** [comment]
I had an idea. Instead of having a graph layout, how about we have a generic `TypeLayout`?
Suppose I need to distribute this an object of this type:
```julia
type complex
   f1::Vector
   f2::AbstractArray{Any,2}
end
```
We can have then have a TypeLayout that describes the layouts of `f1` and `f2`:

```julia
type TypeLayout{T<:Layout}
   layouts::Vector{T}
end

obj = complex(ones(10), ones(10,10))
distribute(obj, TypeLayout([RowLayout(), ColumnLayout()]))
```
Implementing `slice` and `gather` should then be easy and similar to what I've written in `graph-layouts.jl`.

**Comment by pranavtbhat** [comment]
This commit contains the generic TypeLayout definition as well as generic slice and gather methods. 

Can we also have a SingletonLayout? Sometimes I need to pass a single value into a map (a common argument). For example:
```julia
multiply_by(x,y) = x*y
result = gather(Context(), map(multiply_by, distribute(rand(10)), distribute(5, SingletonLayout{Int}())))
```
This layout will simply produce an array of the same element, and slice it.

File: src/ComputeFramework.jl [hunk 4]
```diff
@@ -76,6 +76,8 @@ include("file-nodes.jl")
 include("redistribute.jl")
 include("sort.jl")
 
+include("graph-layouts.jl")
```
Comment (shashi) [inline src/ComputeFramework.jl]: File doesn't exist anymore.

**Comment by shashi** [comment]
> Can we also have a SingletonLayout?

how is it different from `distribute` with `Bcast` layout?

File: src/ComputeFramework.jl (hunk 4 above)
Comment (pranavtbhat) [inline src/ComputeFramework.jl]: Oops sorry.

**Comment by pranavtbhat** [comment]
Oh I hadn't checked BroadcastLayout. Never mind its the same thing.

**Comment by pranavtbhat** [comment]
I guess metadata solves this problem.
